File: src/app/wasm/fibonacci/fibonacci.ts
export const fibonacciLoop = (num: number) => {
  let a = 1;
  let b = 0;
  let temp: number;

  while (num >= 0) {
    temp = a;
    a = a + b;
    b = temp;
    num--;
  }

  return b;
};

export const fibonacciRec = (num: number) => {
  if (num <= 1) {
    return 1;
  }

  return fibonacciRec(num - 1) + fibonacciRec(num - 2);
};

export const fibonacciMemo = (num: number, memo?: Record<number, number>) => {
  memo = memo || {};

  if (memo[num]) {
    return memo[num];
  }

  if (num <= 1) {
    return 1;
  }

  return (memo[num] = fibonacciMemo(num - 1, memo) + fibonacciMemo(num - 2, memo));
};

File: src/app/wasm/fibonacci/fibonacci.c
#include <emscripten.h>

EMSCRIPTEN_KEEPALIVE
int fibonacciLoop(int num) {
  int a = 1;
  int b = 1;

  while (num-- > 1) {
    int t = a;
    a = b;
    b += t;
  }
  return b;
}

EMSCRIPTEN_KEEPALIVE
int fibonacciRec(int num) {
  if (num <= 1)
    return 1;

  return fibonacciRec(num - 1) + fibonacciRec(num - 2);
}

int memo[10000];

EMSCRIPTEN_KEEPALIVE
int fibonacciMemo(int num) {
  if (memo[num] != -1)
    return memo[num];

  if (num == 1 || num == 2) {
    return 1;
  } else {
    return memo[num] = fibonacciMemo(num - 1) + fibonacciMemo(num - 2);
  }
}